File: code/ch05/5.1.3-path-sum-II.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();

        if(root == null) return res;

        Deque<Integer> path = new ArrayDeque<>();
        pathSum(root, sum, path, res);

        return res;
    }

    public void pathSum(TreeNode node, int sum, Deque<Integer> path, List<List<Integer>> res) {
        if (node == null)  return;

        // 沿途结点必须选择，这个时候要做两件事：1、sum 减去这个结点的值；2、添加到 path 里
        sum -= node.val;
        path.addLast(node.val);

        if (sum == 0 && node.left == null && node.right == null) {
            // path 全局只有一份，必须做拷贝
            res.add(new ArrayList<>(path));
            // 注意：这里 return 之前必须重置
            path.removeLast();
            return;
        }

        pathSum(node.left, sum, path, res);
        pathSum(node.right, sum, path, res);
        // 递归完成以后，必须重置变量
        path.removeLast();      
    }
}

File: code/ch05/5.1.3-path-sum-II.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        def helper(root: TreeNode, sum: int, path: List):
            if not root:
                return
            if not root.left and not root.right and sum - root.val == 0:
                path += [root.val]
                ans.append(path)
            helper(root.left, sum - root.val, path + [root.val])
            helper(root.right, sum - root.val, path + [root.val])

        ans = []
        helper(root, sum, [])
        return ans
